### ISSUE TITLE
LAN rules are not really applying in the background unless you disable the firewall/purge rules and then re-enable

### ISSUE THREAD

**Issue by est61** [ukanth/afwall#1031]
Say for example you want to ONLY allow LAN & VPN requests through Firefox.

**To Reproduce:**
1. Allow LAN & VPN in AFWall for now.
2. Keep VPN App disconnected, because we are testing only the LAN from Firefox.
3. Go to your local router IP in Firefox to see if it loads -> result: (it wont load)
4. Go back to AFWall, and disable the firewall -> result: (error purging iptables)
5. Go to 3dot menu and click > Save
6. Go to 3dot menu and click > Apply
7. Go to 3dot menu and click > Disable the firewall -> result: (rules deleted successfully) - if you still get Step 4 error here, Repeat step 6 one more time.
8. Go to 3dot menu and click > Enable Firewall again.
9. Go to your local router IP in Firefox to see if it loads -> result: (now it loads the local LAN IP)

To Disallow this you need to repeat these steps again.

**Expected Result:**
Firstly Step 5,6 & 7 should be unecessary.  This is separate bug.
If you do not do Step 5 & 6 you will keep getting Step 4 error.

Secondly Step 4-8 should be without a doubt unecessary to enable the lan.  Im under the impression that this worked on version 3.1.0, although i did not test this.  I only noticed this now.

Bear in mind that under Preferences > Experimental > Enable Inbound Connections is UNCHECKED.

Basically im reporting 2 bugs here:

The first bug occurs when trying to Disable or Delete the Rules, you need to follow steps 4-7 to actually delete the rules properly, if it still shows error, then you have to repeat Step 6 right after Step 6 again.

The second bug is that the LAN rules dont get activated in the background unless disabling the firewall and re-enabling it.

Version: 3.2.0
Android Pie
Oneplus 6T

**Comment by est61** [comment]
It seems as though for rules to be deleted successfully on first go you need to click Apply twice.  By doing this it seems that it only really Applies ALL rules after the second time applying.  Very strange.

Ive reproduced this multiple times.

**Comment by AbsurdlySuspicious** [comment]
Same happens to me too, but affects all permissions, including wifi and mobile. LOS 9.0

**Comment by varhub** [comment]
Similar issue there (xiaomi Note 7 / MIUI 10.2)
Since 3.2.0 and "dismiss" feature apply-rules success ratio is about 40%. It seems like a race condition due most times toasts messages "error applying" and "success" both appears about progress 98%-105% and real success only depends on messages sorting and interval between them.
Also, when rules fails to apply, there are few seconds that 3dot menu shows "save" instead "apply" so one must wait to try to reapply.

**Comment by h1z1** [comment]
Something to check is the output from iptables-save.  I noticed something setting

```
# iptables-save > ipt.txt
# grep quota -n ipt.txt
234:-A bw_INPUT -m quota2 ! --name globalAlert  --quota 2097152 
239:-A bw_OUTPUT -m quota2 ! --name globalAlert  --quota 2097152 
```
.. which is bogus.

```
# iptables-restore  < ipt.txt
iptables-restore v1.6.1: quota2: option "--name" cannot be inverted.

Error occurred at line: 234
Try `iptables-restore -h' or 'iptables-restore --help' for more information.
# 
```
I don't know what is setting it.

```
# sed -e 's/quota2 !/quota2 /g' -i ipt.txt
# grep quota -n ipt.txt
234:-A bw_INPUT -m quota2  --name globalAlert  --quota 2097152 
239:-A bw_OUTPUT -m quota2  --name globalAlert  --quota 2097152 
# iptables-restore  < ipt.txt
# 
```

I take that back, it's coming from /system/bin/netd

Plot thickens

```
# which iptables
/system/bin/iptables
# 
# find / -type f -name iptables 2>/dev/null |while read f; do md5sum $f; $f -v 2>&1 |head -1; done
0785cfef031357718df33457ff84b2af  /system/bin/iptables
iptables v1.6.1: no command specified
a200373e3c2a4daea76e43dba3b0badc  /sbin/.magisk/mirror/data/data/dev.ukanth.ufirewall/app_bin/iptables
iptables v1.4.20: no command specified
0785cfef031357718df33457ff84b2af  /sbin/.magisk/mirror/system/bin/iptables
iptables v1.6.1: no command specified
a200373e3c2a4daea76e43dba3b0badc  /data/data/dev.ukanth.ufirewall/app_bin/iptables
iptables v1.4.20: no command specified
```

Much older version of iptables.

**Comment by id64350970** [comment]
Same problem on Galaxy A50 running Android 9 rooted with Magisk 20.1. I have the message "error purging iptables" when disabling the firewall. The rules are also not applied at boot and every apps can access internet. I need to click Apply and enable sometimes up to 12 times until it finally work. Attached is the log.
[IPv4rules.log](https://github.com/ukanth/afwall/files/3812830/IPv4rules.log)

**Comment by h1z1** [comment]
Does changing "Startup Delay" under Preferences->Experimental, make any difference?  

The rules themselves could certainly be better but they should be getting applied on boot.

**Comment by id64350970** [comment]
@h1z1 It still does not apply the rules when I enable the Startup Delay function. I think the problem that cause the 'error purging iptables' is also preventing the rules to be applied at boot.

**Comment by ukanth** [comment]
Please share the rules with log messages from menu-show rules-export

**Comment by id64350970** [comment]
@ukanth Following is the log with Startup Delay enabled. You can find the log without the function enabled in one of my previous message. The version 3.1.0 is working fine. I'm using it until the solution is found. The app has been updated to 3.2.0 via Google Play for this test.
[IPv4rules_SDelay.log](https://github.com/ukanth/afwall/files/3837562/IPv4rules_SDelay.log)

**Comment by ukanth** [comment]
Fixed in 3.3.0. Closing.
